src/agent/coverage/src/block.rs

```diff
@@ -119,6 +123,99 @@ impl BlockCov {
     }
 }
 
+mod array {
+    use std::collections::BTreeMap;
+    use std::fmt;
+
+    use serde::de::{self, Deserializer, Visitor};
+    use serde::ser::{SerializeSeq, Serializer};
+
+    use super::BlockCov;
+
+    type BlockCovMap = BTreeMap<u64, BlockCov>;
+
+    pub fn serialize<S>(data: &BlockCovMap, ser: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let mut seq = ser.serialize_seq(Some(data.len()))?;
+        for (_, v) in data {
+            seq.serialize_element(v)?;
+        }
+        seq.end()
+    }
+
+    pub fn deserialize<'d, D>(de: D) -> Result<BlockCovMap, D::Error>
+    where
+        D: Deserializer<'d>,
+    {
+        de.deserialize_seq(FlattenVisitor)
+    }
+
+    struct FlattenVisitor;
+
+    impl<'d> Visitor<'d> for FlattenVisitor {
+        type Value = BlockCovMap;
+
+        fn expecting(&self, f: &mut fmt::Formatter) -> fmt::Result {
+            write!(f, "array of blocks")
+        }
+
+        fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+        where
+            A: de::SeqAccess<'d>,
+        {
+            let mut map = Self::Value::default();
+
+            while let Some(block) = seq.next_element::<BlockCov>()? {
+                map.insert(block.offset, block);
+            }
+
+            Ok(map)
+        }
+    }
+}
+
+mod hex {
+    use std::fmt;
+
+    use serde::de::{self, Deserializer, Visitor};
+    use serde::ser::Serializer;
+
+    pub fn serialize<S>(data: &u64, ser: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        let s = format!("0x{:x}", data);
```

Comment:
This isn't supported by python's json parser.  We shouldn't use a serialization format that requires custom parsers. 

Example:

```
>>> json.loads('{"a": 3}')
{'a': 3}
>>> json.loads('{"a": 0x3}')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/usr/lib/python3.8/json/__init__.py", line 357, in loads
    return _default_decoder.decode(s)
  File "/usr/lib/python3.8/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/usr/lib/python3.8/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
json.decoder.JSONDecodeError: Expecting ',' delimiter: line 1 column 8 (char 7)```

Reply:
A string hex value is specifically suboptimal, as you can't immediately use it.

Reply:
It doesn't get serialized as a prefix numeric literal, it gets serialized as a string (in double quotes).

Reply:
Understood.

That's what I meant by a string hex value can't be immediately used.  It means we have to write a custom parser to use the data.

"0x20" isn't a value that users can immediately use.  They have to parse the string as hex, then use that.

Reply:
Ah. Yeah, that's true if we want to optimize for doing arithmetic.

There are two problems I'm trying to solve here:
1. JSON numbers _can_ be lossy (e.g. if they pass through a JS interpreter)
2. We are talking about memory locations and offsets. I assume users will typically want to _display_ those addrs/offsets, rather than do arithmetic on them. In shell one-liner, it'd be annoying to format a decimal number as hex. So, I'm also optimizing for a different use case.

Reply:
To be clear on my ambient goal: make it trivial to get modoffset (with hex offsets) from the JSON format.

We shouldn't ever run into `Number.MAX_SAFE_INTEGER` in practice with these _offsets_. Arguably we should require that they be `u32` values in the first place. I'd feel more strongly about it if we intended to represent arbitrary (including possibly invalid) 64-bit addresses.

Reply:
Per discussion offline: for this format, we'll switch to serializing the offsets as `Number` values, since we control their generation, and they are _really_ `u32` values. Will change and add comments to that effect.